cmd/dataexpl/retr_fil.go

```diff
@@ -170,168 +157,69 @@ const (
-	resCh := make(chan error, 1)
-	var resSent bool
+	request := types2.RetrievalRequest{
+		RetrievalID: rid,
+		Cid:         file,
+		Selector:    rsn,
+		LinkSystem:  linkSystem,
+	}
 
-	var retrCode = RetrResSuccess
+	request.PreloadLinkSystem = cidlink.DefaultLinkSystem()
+	request.PreloadLinkSystem.SetReadStorage(uselessWrapperStore)
+	request.PreloadLinkSystem.SetWriteStorage(uselessWrapperStore)
+	request.PreloadLinkSystem.TrustedStorage = true
 
-	var data int64
+	ctx, cancel := context.WithCancel(ctx)
 
 	go func() {
-		defer func() {
-			h.filRetrPs.Unsub(subscribeEvents, "events")
-			if retrDone != nil {
-				retrDone()
-			}
-		}()
-		defer cancel()
-
-		for {
-			var evt lapi.RetrievalInfo
-			select {
-			case <-ctx.Done():
-				if !resSent {
-					go func() {
-						err := h.api.ClientCancelRetrievalDeal(context.Background(), retrievalRes.DealID)
-						if err != nil {
-							log.Errorw("cancelling deal failed", "error", err)
-						}
-					}()
-				}
-
-				retrCode = RetrResRetrievalTimeout
-				resCh <- xerrors.New("Retrieval Timed Out")
-
-				log.Infow("retrieval done", "reason", "timed out", "duration", time.Since(start))
-				return
-			case evti := <-subscribeEvents:
-				evt = evti.(lapi.RetrievalInfo)
-				if evt.ID != retrievalRes.DealID {
-					// we can't check the deal ID ahead of time because:
-					// 1. We need to subscribe before retrieving.
-					// 2. We won't know the deal ID until after retrieving.
-					continue
-				}
-			}
-
-			event := "New"
-			if evt.Event != nil {
-				event = retrievalmarket.ClientEvents[*evt.Event]
-			}
-
-			atomic.StoreInt64(&data, int64(evt.BytesReceived))
-
-			fmt.Printf("Recv %s, Paid %s, %s (%s), %s\n",
-				types.SizeStr(types.NewInt(evt.BytesReceived)),
-				types.FIL(evt.TotalPaid),
-				strings.TrimPrefix(event, "ClientEvent"),
-				strings.TrimPrefix(retrievalmarket.DealStatuses[evt.Status], "DealStatus"),
-				time.Now().Sub(start).Truncate(time.Millisecond),
-			)
-
-			switch evt.Status {
-			case retrievalmarket.DealStatusOngoing:
-				if apiStore != nil && !resSent {
-					resCh <- nil
-					resSent = true
-				}
-			case retrievalmarket.DealStatusCompleted:
-				if !resSent {
-					resCh <- nil
-				}
-				log.Infow("retrieval done", "reason", "completed", "duration", time.Since(start))
-				return
-			case retrievalmarket.DealStatusRejected:
-				if !resSent {
-					retrCode = RetrResRetrievalRejected
-					resCh <- xerrors.Errorf("Retrieval Proposal Rejected: %s", evt.Message)
-				}
-				log.Infow("retrieval done", "reason", "rejected", "duration", time.Since(start))
-				return
-			case
-				retrievalmarket.DealStatusDealNotFound,
-				retrievalmarket.DealStatusErrored:
-				if !resSent {
-					retrCode = RetrResRetrievalErr
-					resCh <- xerrors.Errorf("Retrieval Error: %s", evt.Message)
-				}
-				log.Infow("retrieval done", "reason", "errored/not found", "duration", time.Since(start))
-				return
-			}
+		stats, err := lassie.Fetch(ctx, request)
```

Comment:
Fetch happens here, in a goroutine, so that files can be read as blocks are downloaded (thanks to the magical ReadBlockBlockstore)